vcs/vcs.go

```diff
@@ -18,63 +16,61 @@ var (
 	ErrNoNearestVCS = errors.New("could not find nearest VCS repository in directory")
 )
 
-// VCSIn returns the type of VCS repository rooted at a directory, or
+// In returns the type of VCS repository rooted at a directory, or
 // ErrNoVCSInDir if none is found.
-func vcsIn(dirname string) (cli.VCS, error) {
-	either := monad.EitherVCS{}
-	result := either.
-		BindVCS(findVCSFolder(cli.Git, filepath.Join(dirname, ".git"))).
-		BindVCS(findVCSFolder(cli.Subversion, filepath.Join(dirname, ".svn"))).
-		BindVCS(findVCSFolder(cli.Mercurial, filepath.Join(dirname, ".hg"))).
-		BindVCS(findVCSFolder(cli.Bazaar, filepath.Join(dirname, ".bzr")))
-	if result.Err != nil {
-		return 0, result.Err
-	}
-	if result.Result == 0 {
-		return 0, ErrNoVCSInDir
+func In(dirname string) (VCS, error) {
+	for _, vcs := range Types {
+		ok, err := files.ExistsFolder(filepath.Join(dirname, MetadataFolder(vcs)))
+		if err != nil {
+			return 0, err
+		}
+		if ok {
+			return vcs, nil
+		}
 	}
-	return result.Result, nil
+	return 0, ErrNoVCSInDir
 }
 
-// NearestVCS returns the type and directory of the nearest VCS repository
+// Nearest returns the type and directory of the nearest VCS repository
 // containing the directory, or ErrNoNearestVCS if none is found.
-func NearestVCS(dirname string) (vcsType cli.VCS, vcsDir string, err error) {
```

Comment:
Why the move away from named returns?

Reply:
I just avoid them where possible, because assignment in Go follows unintuitive shadowing rules. I usually only use named returns when returning multiple values of the same type.

Example of weird shadowing:
```go
func Foo() (bar string) {
	bar, err := Baz() // This is a _different_ bar, shadowing the original
	return            // Implicitly returns `bar = ""`, the string type's zero value
}
```

- [Documentation on named returns](https://tour.golang.org/basics/7)
- More shadowing examples:
	- [Go 2 proposal removing shadowing](https://github.com/golang/go/issues/21114)
	- [Blog post](https://blog.charmes.net/post/scope-shadowing-go/)
	- [Mailing list thread](https://groups.google.com/forum/#!topic/golang-nuts/VeJTB9dhhK4)